homeassistant/components/fritzbox/__init__.py

```diff
@@ -11,80 +11,91 @@
 
-ATTR_STATE_BATTERY_LOW = "battery_low"
-ATTR_STATE_DEVICE_LOCKED = "device_locked"
-ATTR_STATE_HOLIDAY_MODE = "holiday_mode"
-ATTR_STATE_LOCKED = "locked"
-ATTR_STATE_SUMMER_MODE = "summer_mode"
-ATTR_STATE_WINDOW_OPEN = "window_open"
+def ensure_unique_hosts(value):
+    """Validate that all configs have a unique host."""
+    vol.Schema(vol.Unique("duplicate host entries found"))(
+        [socket.gethostbyname(entry[CONF_HOST]) for entry in value]
+    )
+    return value
 
 
 CONFIG_SCHEMA = vol.Schema(
     {
-        DOMAIN: vol.Schema(
-            {
-                vol.Required(CONF_DEVICES): vol.All(
-                    cv.ensure_list,
-                    [
-                        vol.Schema(
-                            {
-                                vol.Required(CONF_HOST): cv.string,
-                                vol.Required(CONF_PASSWORD): cv.string,
-                                vol.Required(CONF_USERNAME): cv.string,
-                            }
-                        )
-                    ],
+        DOMAIN: vol.All(
+            cv.ensure_list,
+            [
+                vol.Schema(
+                    {
+                        vol.Required(CONF_HOST, default=DEFAULT_HOST): cv.string,
+                        vol.Required(CONF_PASSWORD): cv.string,
+                        vol.Required(
+                            CONF_USERNAME, default=DEFAULT_USERNAME
+                        ): cv.string,
+                    }
                 )
-            }
+            ],
+            ensure_unique_hosts,
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
 
-def setup(hass, config):
-    """Set up the fritzbox component."""
-
-    fritz_list = []
+async def async_setup(hass, config):
+    """Set up the AVM Fritz!Box integration."""
+    if DOMAIN in config:
+        for entry_config in config[DOMAIN]:
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN, context={"source": "import"}, data=entry_config
+                )
+            )
 
-    configured_devices = config[DOMAIN].get(CONF_DEVICES)
-    for device in configured_devices:
-        host = device.get(CONF_HOST)
-        username = device.get(CONF_USERNAME)
-        password = device.get(CONF_PASSWORD)
-        fritzbox = Fritzhome(host=host, user=username, password=password)
-        try:
-            fritzbox.login()
-            _LOGGER.info("Connected to device %s", device)
-        except LoginError:
-            _LOGGER.warning("Login to Fritz!Box %s as %s failed", host, username)
-            continue
+    def logout_fritzboxes(event):
+        """Close all connections to the fritzboxes."""
+        if DOMAIN in hass.data and CONF_CONNECTIONS in hass.data[DOMAIN]:
+            for fritz in hass.data[DOMAIN][CONF_CONNECTIONS]:
+                fritz.logout()
 
-        fritz_list.append(fritzbox)
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, logout_fritzboxes)
```

Comment:
Move this to the `async_entry_setup` and register one listener per entry. The callback should just logout from the fritz instance for the corresponding config entry.

Reply:
How to pass `entry.entry_id` to the listener?

Reply:
Or is this working with local `logout_fritzboxes` method?

Reply:
The callback function will remember its scope, yes. So as long as you have access to the fritz instance in the callback, that's enough.